<a href="https://colab.research.google.com/github/Joaalbmed/IA-y-BD/blob/main/Exploraci%C3%B3n_de_datos_(EDA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Q1. Pandas version

What version of Pandas did you install?

You can get the version information using the __version__ field:

pd.__version__


Getting the data

For this homework, we'll use the Car Fuel Efficiency dataset. Download it from here.

You can do it with wget:

wget https://raw.githubusercontent.com/bigdatadatafan/datasets-clase/main/car_fuel_efficiency.csv



Or just open it with your browser and click "Save as...".

Now read it with Pandas.

Q2. Records count

How many records are in the dataset?

    4704
    8704
    9704
    17704

Q3. Fuel types

How many fuel types are presented in the dataset?

    1
    2
    3
    4

Q4. Missing values

How many columns in the dataset have missing values?

    0
    1
    2
    3
    4

Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

    13.75
    23.75
    33.75
    43.75

Q6. Median value of horsepower

    Find the median value of the horsepower column in the dataset.
    Next, calculate the most frequent value of the same horsepower column.
    Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
    Now, calculate the median value of horsepower once again.

Has it changed?

    Yes, it increased
    Yes, it decreased
    No

Q7. Sum of weights

    Select all the cars from Asia
    Select only columns vehicle_weight and model_year
    Select the first 7 values
    Get the underlying NumPy array. Let's call it X.
    Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
    Invert XTX.
    Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
    Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
    What's the sum of all the elements of the result?

    Note: You just implemented linear regression. We'll talk about it in the next lesson.

    0.051
    0.51
    5.1
    51


In [ ]:

!pip install pandas
import pandas as pd

In [ ]:
pd.__version__

'2.2.2'

In [ ]:
!wget https://raw.githubusercontent.com/bigdatadatafan/datasets-clase/main/car_fuel_efficiency.csv

--2025-10-15 18:04:28--  https://raw.githubusercontent.com/bigdatadatafan/datasets-clase/main/car_fuel_efficiency.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 874188 (854K) [text/plain]
Saving to: ‘car_fuel_efficiency.csv.2’

car_fuel_efficiency 100%[===================>] 853.70K  --.-KB/s    in 0.01s   

2025-10-15 18:04:28 (82.4 MB/s) - ‘car_fuel_efficiency.csv.2’ saved [874188/874188]



# Task
Write Python code to answer the questions in the text block using the data from the "car_fuel_efficiency.csv" file.

## Load the data

### Subtask:
Read the `car_fuel_efficiency.csv` file into a pandas DataFrame.


**Reasoning**:
Read the CSV file into a pandas DataFrame as instructed.



In [ ]:
df = pd.read_csv('car_fuel_efficiency.csv')

## Answer q2

### Subtask:
Determine the number of records in the dataset.


**Reasoning**:
Get the number of rows in the DataFrame `df`.



In [ ]:
len(df)

9704

## Answer q3

### Subtask:
Identify and count the unique fuel types in the dataset.


**Reasoning**:
Select the 'fuel_type' column and count the number of unique values using nunique().



In [ ]:
num_fuel_types = df['fuel_type'].nunique()
print(num_fuel_types)

2


## Answer q4

### Subtask:
Find and count the columns that contain missing values.


**Reasoning**:
Use isnull() and sum() to count missing values per column, then filter for columns with more than 0 missing values and count them.



In [ ]:
missing_values = df.isnull().sum()
columns_with_missing_values = missing_values[missing_values > 0]
num_columns_with_missing_values = len(columns_with_missing_values)
print(num_columns_with_missing_values)

4


## Answer q5

### Subtask:
Calculate the maximum fuel efficiency for cars from Asia.


**Reasoning**:
Filter the DataFrame for cars from Asia and calculate the maximum fuel efficiency.



In [ ]:
asia_cars = df[df['origin'] == 'Asia']
max_fuel_efficiency_asia = asia_cars['fuel_efficiency_mpg'].max()
print(max_fuel_efficiency_asia)

23.759122836520497


## Answer q6

### Subtask:
Calculate and compare the median horsepower before and after filling missing values with the mode.


**Reasoning**:
Calculate the median and mode of the 'horsepower' column, fill missing values with the mode, calculate the new median, and compare the two medians.



In [ ]:
median_horsepower_before = df['horsepower'].median()
mode_horsepower = df['horsepower'].mode()[0]
df['horsepower'] = df['horsepower'].fillna(mode_horsepower)
median_horsepower_after = df['horsepower'].median()

if median_horsepower_after > median_horsepower_before:
    change = "Yes, it increased"
elif median_horsepower_after < median_horsepower_before:
    change = "Yes, it decreased"
else:
    change = "No"

print(f"Median horsepower before filling missing values: {median_horsepower_before}")
print(f"Mode of horsepower: {mode_horsepower}")
print(f"Median horsepower after filling missing values: {median_horsepower_after}")
print(f"Has the median changed? {change}")

Median horsepower before filling missing values: 149.0
Mode of horsepower: 152.0
Median horsepower after filling missing values: 152.0
Has the median changed? Yes, it increased


## Answer q7

### Subtask:
Perform the specified linear regression steps for cars from Asia using the `vehicle_weight` and `model_year` columns and calculate the sum of the elements in the resulting vector `w`.


**Reasoning**:
Perform the linear regression steps and calculate the sum of the elements in the resulting vector w.



In [ ]:
import numpy as np

asia_cars_q7 = df[df['origin'] == 'Asia']
X_df = asia_cars_q7[['vehicle_weight', 'model_year']]
X = X_df.head(7).to_numpy()
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
XT = X.T
XTX = XT @ X
XTX_inv = np.linalg.inv(XTX)
intermediate_result = XTX_inv @ XT
w = intermediate_result @ y
sum_w = np.sum(w)
print(sum_w)

0.5187709081074016


## Summary:

### Q&A

1.  **Number of records in the dataset:** There are 9704 records in the dataset.
2.  **Unique fuel types:** There are 2 unique fuel types in the dataset.
3.  **Columns with missing values:** There are 4 columns with missing values.
4.  **Maximum fuel efficiency for cars from Asia:** The maximum fuel efficiency for cars from Asia is approximately 23.76 MPG.
5.  **Change in median horsepower after filling missing values with the mode:** The median horsepower increased from 149.0 to 152.0 after filling missing values with the mode.
6.  **Sum of the elements in the resulting vector `w` from the linear regression:** The sum of the elements in the vector `w` is approximately 0.5188.

### Data Analysis Key Findings

*   The dataset contains 9704 records.
*   There are 2 unique fuel types present in the dataset.
*   Four columns in the dataset contain missing values.
*   The maximum fuel efficiency observed for cars originating from Asia is 23.76 MPG.
*   Imputing missing values in the 'horsepower' column with the mode (152.0) resulted in an increase in the median horsepower from 149.0 to 152.0.
*   A linear regression performed on a subset of Asian cars using 'vehicle\_weight' and 'model\_year' resulted in a weight vector `w` whose elements sum to approximately 0.5188.

### Insights or Next Steps

*   Investigate the columns with missing values to determine the best strategy for handling them (e.g., imputation, removal) based on the nature of the data and the analysis goals.
*   Explore the characteristics of the two unique fuel types and their distribution within the dataset.
